onnxruntime/core/session/inference_session.h

```diff
@@ -466,10 +455,21 @@ class InferenceSession {
   std::unique_ptr<SessionState> session_state_;
 
  private:
-  // Threadpool for this session
+  // Threadpools per session. These are initialized and used for the entire duration of the session
+  // when use_per_session_threads is true.
   std::unique_ptr<onnxruntime::concurrency::ThreadPool> thread_pool_;
   std::unique_ptr<onnxruntime::concurrency::ThreadPool> inter_op_thread_pool_;
 
+  // Global threadpools. These are intialized and used when use_per_session_threads is false *and*
+  // the environment is created with create_global_thread_pools = true.
+  onnxruntime::concurrency::ThreadPool* intra_op_thread_pool_from_env_{};
+  onnxruntime::concurrency::ThreadPool* inter_op_thread_pool_from_env_{};
+
+  // initialized from session options
+  // Determines which threadpools will be intialized and used for the duration of this session.
+  // If true, use the per session ones, or else the global threadpools.
+  bool use_per_session_threads_;
+
```

Comment:
Would it be simpler to have something like:

  onnxruntime::concurrency::ThreadPool* intra_op_thread_pool_{};
  onnxruntime::concurrency::ThreadPool* inter_op_thread_pool_{};
  std::unique_ptr<onnxruntime::concurrency::ThreadPool> owned_intra_op_thread_pool_;
  std::unique_ptr<onnxruntime::concurrency::ThreadPool> owned_inter_op_thread_pool_;

Create threadpools when needed. Populate the pointers from either the env or the owned unique ptrs based on the session options. From that point on, everything else in InferenceSession can just use the ThreadPool* members.

Reply:
When the env threadpools are in use, the owned and the raw ptrs won't have a 1-1 mapping since the owned unique_ptrs will be null in the session (since the env threadpools are owned by the env). This creates a bit of confusion as well. What I can do is provide a method in the InferenceSession class that returns the correct threadpool raw ptrs based on session options; something like ```pair<ThreadPool*,ThreadPool*> GetThreadPools(const SessionOptions&)``` so that other methods don't have to rely on the use_per_session_threads boolean flag.

Reply:
You could just have a generic std::vector<unique_ptr<ThreadPool>> for threadpools to cleanup to there's less chance someone will try and use them

---
In reply to: [393471791](https://github.com/microsoft/onnxruntime/pull/3177#discussion_r393471791) [](ancestors = 393471791)

Reply:
Resolved by adding 2 new APIs in the session that return the threadpools to use.